toolchain: add support for GCC 15.1 and fix assorted packages

rsalvaterra opened the pull request:
All patches automatically refreshed.

Also fix all packages required by my specific configurations. Any other fixes can be incremental, as separate pull requests/patches.

Signed-off-by: Rui Salvaterra \<rsalvaterra@gmail.com\>

rsalvaterra committed:
toolchain: add support for GCC 15.1

All patches automatically refreshed.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18600
Signed-off-by: Robert Marko <robimarko@gmail.com>

rsalvaterra committed:
mbedtls: fix building with GCC 15.1

Disable -Wunterminated-string-initialization for now.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18600
Signed-off-by: Robert Marko <robimarko@gmail.com>

rsalvaterra committed:
ppp: fix building with GCC 15.1

Backport two patches in order to fix it.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18600
Signed-off-by: Robert Marko <robimarko@gmail.com>

rsalvaterra committed:
linux-atm: fix building with GCC 15.1

Add a patch in order to fix it.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18600
Signed-off-by: Robert Marko <robimarko@gmail.com>

schuettecarsten commented:
`libsepol` fails to build with GCC 15.1

schuettecarsten commented:
`mbedtls` fails to build with GCC 15.1, tons of errors like this:

```
library/ssl_tls13_keys.h:14:40: error: initializer-string for array of 'unsigned char' truncates NUL terminator but destination lacks 'nonstring' attribute (9 chars into 8 available) [-Werror=unterminated-string-initialization]
   14 |     MBEDTLS_SSL_TLS1_3_LABEL(finished, "finished") \
      |                                        ^~~~~~~~~~
library/ssl_tls13_keys.c:38:13: note: in definition of macro 'MBEDTLS_SSL_TLS1_3_LABEL'
   38 |     .name = string,
      |             ^~~~~~
library/ssl_tls13_keys.c:44:5: note: in expansion of macro 'MBEDTLS_SSL_TLS1_3_LABEL_LIST'
   44 |     MBEDTLS_SSL_TLS1_3_LABEL_LIST
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

zheshifandian commented:
> `libsepol` fails to build with GCC 15.1, updating libsepol to 3.8.1 fixes this.
Add 

TARGET_CFLAGS += $(FPIC) -std=gnu17
TARGET_LDFLAGS += $(FPIC)

before MAKE_FLAGS += \

schuettecarsten commented:
The PR needs to be updated after https://github.com/openwrt/openwrt/commit/5c8301e80032b17cc1f3e1f923f3198bb673dea0

rockdrilla commented:
@rsalvaterra PTAL #18647 - here's another pending patch.

rsalvaterra committed:
mtd: fix building with GCC 15.1

Convert old style declarations to modern ones.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18600
Signed-off-by: Robert Marko <robimarko@gmail.com>

hardfalcon commented:
At least kernel 6.6.89 fails to build for MIPS targets such as `ath79/generic` or `realtek/rtl838x` when using GCC 15.1.

My current compile runs haven't finished yet, but it looks like backporting [commit 0f4ae7c6ecb89bfda026d210dcf8216fb67d2333 from `linux-next`](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=0f4ae7c6ecb89bfda026d210dcf8216fb67d2333) solves the issue, and I'm pretty sure that kernel 6.12.26 will require the same backport.

graysky2 commented:
@rsalvaterra - please see https://github.com/openwrt/packages/issues/26446 as you are finalizing this PR. Seems that golang needs something.

rsalvaterra commented:
> @rsalvaterra - please see [openwrt/packages#26446](https://github.com/openwrt/packages/issues/26446) as you are finalizing this PR. Seems that golang needs something.

As written in the first comment, this series fixes strictly my personal tested configurations (which only include mediatek/filogic, x86/64, and octeon, for the time being). I don't use golang at all. Moreover, it belongs to the packages feed, so it's most definitely out of the scope of this series.

That being said, from the output, it seems just a matter of compiling with a C17 dialect, as `bool` became a reserved keyword in C23.

graysky2 commented:
Not sure which file to edit adding the `-std=c17` ... tried:
```
@@ -77,9 +77,9 @@ GO_HOST_TARGET_VARS= \
        CC=gcc \
        CXX=g++ \
        PKG_CONFIG=pkg-config \
-       CGO_CFLAGS="$(HOST_CFLAGS) $(GO_HOST_MORE_CFLAGS)" \
+       CGO_CFLAGS="$(HOST_CFLAGS) $(GO_HOST_MORE_CFLAGS)" -std=c17 \
        CGO_CPPFLAGS="$(HOST_CPPFLAGS) $(GO_HOST_MORE_CPPFLAGS)" \
-       CGO_CXXFLAGS="$(HOST_CFLAGS) $(GO_HOST_MORE_CFLAGS)" \
+       CGO_CXXFLAGS="$(HOST_CFLAGS) $(GO_HOST_MORE_CFLAGS)" -std=c17 \
        CGO_LDFLAGS="$(HOST_LDFLAGS) $(GO_HOST_MORE_LDFLAGS)" \
        GO_GCC_HELPER_CC="$(HOSTCC)" \
        GO_GCC_HELPER_CXX="$(HOSTCXX)" \
diff --git a/libs/libgee/Makefile b/libs/libgee/Makefile
index 965b49db7c94..86f45f62d84e 100644
```

Bit that did not help.

jonasjelonek commented:
Noticed two kernel build regressions while building for Banana Pi R4 (mediatek/filogic, both 6.6 and 6.12). Patches exist somewhat upstream for this. Is this supposed to be included here or in a separate PR?

rsalvaterra commented:
> Noticed two kernel build regressions while building for Banana Pi R4 (mediatek/filogic, both 6.6 and 6.12). Patches exist upstream for this (pending and to-be-backported). Is this supposed to be included here or in a separate PR?

I can include them here, sure. Can you point me to the specific patches?

jonasjelonek commented:
> EDIT: Are they compiler patches? If not, it can be made a separate pull request.

No, just for in-kernel `drivers/net/wireguard` and `drivers/power/supply`. I already prepared two OpenWrt patches/commits for that in my tree [here](https://github.com/jonasjelonek/openwrt/tree/gcc15-fixes). 
But sure, I can do a separate PR for these.

robimarko commented on package/libs/libsepol/Makefile:
This should not be needed, it compiles with GCC14 and 15 just fine for me without forcing GNU17

robimarko commented:
Thanks! Rebased on top of main and merged!
